lib/cyrusdb_twom.c

```diff
@@ -0,0 +1,2597 @@
+#include "xmalloc.h"
+#include "xunlink.h"
+
+#define XXH_STATIC_LINKING_ONLY /* access advanced declarations */
+#define XXH_INLINE_ALL          /* maximum optimise */
+#define XXH_IMPLEMENTATION      /* access definitions */
+#include "xxhash.h"
+
+/********** TUNING *************/
+
+/* don't bother rewriting if the database has less than this much extra */
+#define MINREWRITE 16834
+/* number of skiplist levels - 31 gives us binary search to 2^32 records.
+ * This is the limit to what we can calculate from a single call to random(),
+ * but skiplist had 20, and that was enough
+ * for most real uses.  31 is heaps. */
+#define MAXLEVEL 31
+
+/* release lock in foreach at least every N records */
+#define FOREACH_LOCK_RELEASE 1024
+
+/* format specifics */
+#undef VERSION /* defined in config.h */
+#define VERSION 1
+
+/* type aliases */
+#define LLU long long unsigned int
+#define LU long unsigned int
+
+/* record types */
+#define DUMMY 0
+#define COMMIT 1
+#define ADD 2
+#define FATADD 3
+#define REPLACE 4
+#define FATREPLACE 5
+#define DELETE 6
+#define FATDELETE 7
+const char *typestr[] = { "DUMMY", "COMMIT", "ADD", "FATADD",
+                          "REPLACE", "FATREPLACE", "DELETE", "FATDELETE" };
+uint8_t ptroffset[8]      = { 8,  8,  8, 24, 16, 32, 16, 32 };
+uint8_t ancestoroffset[8] = { 0,  0,  0,  0,  8, 24,  8, 24 };
+uint8_t fatrecord[8]      = { 0,  0,  0,  1,  0,  1,  0,  1 };
+uint8_t haskey[8]         = { 0,  0,  1,  1,  1,  1,  1,  1 };
+uint8_t hasval[8]         = { 0,  0,  1,  1,  1,  1,  0,  0 };
+
+/********** DATA STRUCTURES *************/
+
+struct tm_header {
+    /* header info */
+    unsigned char uuid[16];
+    uint16_t version;
+    uint16_t checksum_engine;
+    uint32_t flags;
+    uint64_t generation;
+    uint64_t num_records;
+    uint64_t num_commits;
+    size_t dirty_size;
+    size_t repack_size;
+    size_t current_size;
+    uint32_t maxlevel;
+};
+
+struct tm_file {
+    // file descriptor
+    int fd;
+    int has_lock;
+    // mmap
+    char *base;
+    size_t size;
+    int dirty;
+    // header info
+    struct tm_header header;
+    uint32_t (*csum)(const char *base, size_t len);
+    int (*compar)(const char *a, size_t alen, const char *b, size_t blen);
+    // tracking
+    int refcount;
+    struct tm_file *next;
+};
+
+/* a location in the twom file.  We always have:
+ * offset: if "is_exactmatch" this points to the record
+ *         with the matching key, otherwise it points to
+ *         the 'compar' order previous record.
+ * backloc: the records that point TO this location
+ *          at each level.
+ * end can be used to see if anything in
+ * the file may have changed and needs relocation.
+ */
+struct tm_loc {
+    struct tm_file *file;
+    size_t end;  // pointers are only valid when end matches txn end
+    // location
+    size_t offset;     // current position
+    size_t backloc[MAXLEVEL+1];
+    int is_exactmatch; // key was passed in; did we match it?
```

Comment:
Might be advantageous to put `is_exactmatch` next to `offset` so they're more likely to be in the same cache line?   I get the impression from the comment that they'll typically be accessed together, but with `backloc[]` in between, they'll be 8 * 32 bytes away from each other in memory.

Reply:
Good point, moved them